feat: Verify genesis every time starting

keroro520 opened the pull request:
* Verify genesis block by `BlockVerifier` every time starting node
* Fix capacity calculation and inputs of genesis block
* Verify occupied_capacity of celbase transactions 
* Refactor `tests/uncle_verifier.rs` and add several tests
* **BREAKING CHANGE**: Update genesis hash

keroro520 commented on spec/src/lib.rs:
```diff
@@ -247,12 +256,6 @@ pub mod test {
-            format!("{:x}", tx.hash()),
-            "48168c5b2460bfa698f60e67f08df5298b1d43b2da7939a219ffd863e1380d11"
-        );
-
```
This test makes no sense.

doitian replied:
This is for developers who need to hard-coded `always_success` hash. However, it is going to be removed in #562 .

keroro520 commented on util/app-config/src/lib.rs:
```diff
@@ -217,17 +215,7 @@ fn is_daemon(subcommand_name: &str) -> bool {
-        .verify(consensus.genesis_block())
-        .map_err(Box::new)?;
-    Ok(consensus)
+    spec.to_consensus().map_err(|_| ExitCode::Config)
```
`spec/src/lib.rs` only ensures that genesis is matched with the specs, recorded in `specs/*.toml`.

`src/subcommand/run.rs` do the real block validation (Unable to do this inside `spec/src/lib.rs` cause circular dependency of `ckb_shared` and `ckb-chain-specs`).

quake commented on verification/src/block_verifier.rs:
```diff
@@ -97,6 +96,16 @@ impl<CP: ChainProvider + Clone> CellbaseVerifier<CP> {
             return Err(Error::Cellbase(CellbaseError::InvalidOutput));
         }
 
+        let invalid_capacity = cellbase_transaction.outputs().iter().any(|op| {
```
I think that it's better to add a function to `CellOutput`
```
pub fn is_occupied_capacity_overflow(&self) {
    
}
```

And refactoring this code block to:
```
    if cellbase_transaction.outputs().iter().any(CellOutput::is_occupied_capacity_overflow) {
        return Err(Error::CapacityOverflow);
    }
```

Similar code of `CapacityVerifier` can be refactored to use this function also.

keroro520 replied:
Thanks. Updated and rebased.

keroro520 commented on spec/src/lib.rs:
```diff
@@ -168,19 +169,27 @@ impl ChainSpec {
-                        Capacity::bytes(data.len())
-                            .map(|cap| CellOutput::new(cap, data, Script::default(), None))
-                            .map_err(|err| Box::new(err) as Box<Error>)
+                        let data = data.into_owned();
```
Thanks! It is more clear. Applied and rebased the suggested code.

yangby-cryptape replied:
IMHO, we shouldn't pass a `capacity` to `CellOutput` as an parameter, `capacity` should be computed in `CellOutput::new()`.
(Not in this PR, we can do it later.)

keroro520 commented:
> Need update `test/integration.toml` genesis hash.

Updated.
And added monitor for child process in `test/` https://github.com/nervosnetwork/ckb/pull/582/files#diff-750b6a3c46becef8048e261f521c1d93
